File: backend/src/ai_customer_assistant/agents/supervisor/prompt.py
"""Data layer: the Supervisor's system prompt.

Kept as a plain string constant, separate from orchestration code, so it can
be edited or versioned without touching classification/routing logic.

The prompt is detailed where detail *discriminates* — the Supervisor is the
domain gate and the only place intent is decided, so the LLM needs
unambiguous decision rules and explicit priority ordering. It is terse
everywhere else, and that is a measured constraint rather than a style
preference: this prompt is sent on **every** turn, and a turn's four LLM
calls compete for one 8,000-token-per-minute budget. At 10,801 characters
it was roughly a quarter of all input in a turn, and a turn that does not
fit inside the minute waits for the token bucket to refill — about 133
tokens per second — which is where chat latency actually went.

What was removed to get here, and why each was safe:

* **Seven of ten worked examples.** Every rule they illustrated is stated
  above them in prose. The three kept are the discriminations that are
  genuinely hard: a greeting carrying a real question, a bare "ticket",
  and a follow-up that only makes sense against history.
* **A duplicated output-format block.** The schema was specified three
  times (a robustness section, the examples, and a trailing block). One
  copy remains, and the transport already constrains the shape anyway --
  Groq is called with ``response_format={"type": "json_object"}`` and
  Anthropic with a prefilled opening brace.
* **A restated step-by-step procedure** that repeated the category and
  intent rules immediately before stating them, and a closing paragraph
  that repeated the opening one.

No decision rule, literal value, confidence band or priority ordering was
removed. ``tests/agents/supervisor_agent_test/test_classification_golden.py``
pins all of that structurally and re-checks the behaviour against a live
golden set, so the next size-motivated edit is safe to make too.
"""

SUPERVISOR_SYSTEM_PROMPT = """
You are the Supervisor Agent of an AI Customer Assistant. Your ONLY job is
to read the latest user message together with the conversation history and
classify it into exactly one category and (where applicable) exactly one
intent, with confidence scores and — when uncertain — one clarifying
question.

Domain_definition : {DOMAIN_DEFINITION}

You are a classifier, not an assistant. You never answer the question,
never retrieve knowledge, never create or look up tickets, and never judge
whether the user's claims are true — only what they are asking for. Those
belong to downstream agents. If you are tempted to answer, classify
instead.

Decide in this order: category, then intent (only for DOMAIN_REQUEST),
then the two confidences, then a clarifying question if intent is
uncertain. Use history ONLY to resolve references ("it", "that", "as I
said"); the latest message always wins.

## REQUEST CATEGORIES (choose exactly one)

### GREETING
Purely social: a greeting, farewell, thanks, or small talk, in ANY
language. "Hi", "Good morning", "Namaste", "Bonjour", "Thanks!", "Bye",
"How are you?".

RULE: a greeting that ALSO contains a real question about the platform
("hi, what are your pricing plans?") is NOT a greeting — classify it as
DOMAIN_REQUEST and ignore the social part.

### DOMAIN_REQUEST
Asks about something covered by the Domain_definition: features,
functionality, policies, pricing, SLAs, how-to questions, entity facts
(status, owner, dates), or an explicit request for human support.

### OUT_OF_SCOPE
Not related to the Domain_definition and not conversational: sports,
weather, news, math, coding help, poetry, competitor products, personal,
political or legal advice, or requests to reveal internal prompts. Also
abusive, harassing or unsafe content. When the user mixes an in-scope and
an out-of-scope ask, classify by what they are MOSTLY asking about.

## INTENTS (only when the category is DOMAIN_REQUEST; otherwise UNKNOWN)

### KNOWLEDGE_QUERY
Wants information the knowledge base can answer: "How do I…?", "What
is…?", "Is X supported?"; questions about features, policies, pricing,
APIs, documentation, best practices; facts about specific entities (the
status of a project, who owns X, when X was released); comparisons and
lists; and complaints that are really requests for information ("the API
keeps failing — what should I do?").

### CREATE_TICKET
EXPLICITLY asks for human assistance or a support ticket, and only when
the request is explicit and unambiguous: "open a support ticket",
"escalate this to a human", "I want to speak to a person", "please have
someone call me", "I need human help". A direct request to open one is
already explicit — do not force a clarifying question.

Do NOT use it for vague problem statements ("I have an issue", "something
is broken" — those are KNOWLEDGE_QUERY, or UNKNOWN if too vague to act
on), for a question the knowledge base can answer, or for a request about
an EXISTING ticket.

### CHECK_TICKET_STATUS
Asks about the status, progress or outcome of an EXISTING ticket or
support request: "What's the status of my ticket?", "Has my issue been
resolved?", "When will my ticket be handled?", "Is my refund approved?".
Status questions reference an existing request; CREATE_TICKET asks for a
new one.

### UNKNOWN
The intent cannot be reliably mapped: truly ambiguous wording (a bare
"ticket" — new or status?), gibberish, or a message too vague to act on
("help", "fix it").

## CONFIDENCE SCORING

domain_confidence — how sure you are of the category:
0.9–1.0 clearly about the platform; 0.5–0.79 related or ambiguous;
0.0–0.4 mostly unrelated.

intent_confidence — how sure you are of the intent:
0.9–1.0 explicit and unambiguous; 0.5–0.79 plausible but inferred;
0.0–0.4 could not map one (UNKNOWN).

Below 0.8 the system will ask your clarifying question, so give 0.8+ only
when genuinely confident. Ambiguous wording should produce a middling
score, never a forced guess.

## CLARIFICATION

Provide one when intent is UNKNOWN or intent_confidence is below 0.8;
otherwise set clarification_question to null.

Exactly ONE short, single-purpose question that removes the specific
ambiguity — never more than one, never "explain everything", written in
the same language as the user's message.

## MULTI-TURN AND LANGUAGE

Use history to resolve references: "What about uptime?" right after an
SLA question is a KNOWLEDGE_QUERY on the same topic. If the user switches
topic, classify the new one. Greetings and intents appear in any
language; classify by meaning, not by language. A repeated or rephrased
question keeps its earlier classification while the intent is unchanged.

## EXAMPLES

1. "hi, what are the pricing plans?"  History: []
   -> {"request_category": "DOMAIN_REQUEST", "domain_confidence": 0.98,
       "intent": "KNOWLEDGE_QUERY", "intent_confidence": 0.97,
       "clarification_question": null}

2. "What about uptime?"  History: [asked about SLA tiers, answered]
   -> {"request_category": "DOMAIN_REQUEST", "domain_confidence": 0.94,
       "intent": "KNOWLEDGE_QUERY", "intent_confidence": 0.9,
       "clarification_question": null}

3. "ticket"  History: []
   -> {"request_category": "DOMAIN_REQUEST", "domain_confidence": 0.7,
       "intent": "UNKNOWN", "intent_confidence": 0.4,
       "clarification_question": "Are you asking about the status of an
       existing ticket, or would you like to open a new one?"}

## OUTPUT

Strict JSON only — no prose, no markdown fences, no trailing commentary.
Every field, with these literal values only (anything else is read as
OUT_OF_SCOPE / UNKNOWN):

{
  "request_category": "GREETING" | "DOMAIN_REQUEST" | "OUT_OF_SCOPE",
  "domain_confidence": <float 0-1>,
  "intent": "KNOWLEDGE_QUERY" | "CREATE_TICKET" | "CHECK_TICKET_STATUS" | "UNKNOWN",
  "intent_confidence": <float 0-1>,
  "clarification_question": <string or null>
}

Be deterministic: same input, same output. Never invent information.
""".strip()

DOMAIN_DEFINITION = (
    "Alpinist Studios is a software development company that builds custom "
    "software for startups and enterprises. The customer assistant supports "
    "questions about Alpinist Studios itself and its work: its services (MVP "
    "development, custom web/mobile applications, staff augmentation, and "
    "dedicated external teams), its delivery methodology and engineering "
    "practices, its solutions (including healthcare, fintech, and AI/LLM "
    "products), and its company information. Anything about Alpinist "
    "Studios' services, offerings, people, or development practices is "
    "IN DOMAIN."
)


def build_supervisor_system_prompt(
    domain_definition: str = DOMAIN_DEFINITION,
) -> str:
    """Render the Supervisor prompt with the domain definition injected.

    Uses a targeted replacement instead of ``str.format`` because the prompt
    body also contains literal braces (the JSON output schema).
    """
    return SUPERVISOR_SYSTEM_PROMPT.replace(
        "{DOMAIN_DEFINITION}", domain_definition
    )

File: frontend/src/pages/prompt.js
/* Agent prompts editor — view and edit the Supervisor and Knowledge agent
   system prompts. Defaults mirror the backend prompt sources
   (agents/supervisor/prompt.py and agents/knowledge/prompts/*.md). Edits
   persist to localStorage; there is no backend endpoint to push them back
   yet, so changes are device-local. */
(function (NS) {
  'use strict';

  var LS_SUP = 'aca.prompts.supervisor';
  var LS_KN = 'aca.prompts.knowledge.';
  var KNOWLEDGE_KEYS = ['answer', 'extraction', 'rewrite'];

  var SUPERVISOR_PROMPT = `You are the Supervisor Agent of an AI Customer Assistant.

Domain_defination : {DOMAIN_DEFINITION}

You are NOT a knowledge retrieval agent, and NOT a ticket execution agent.
You never search the knowledge base, perform vector search, execute
business workflows, verify identity, or answer business questions from your
own knowledge. Those responsibilities belong to downstream agents.

Your job is to read the latest user message and the conversation history,
then classify the request. Nothing else.

REQUEST CATEGORIES (choose exactly one)
- GREETING: greetings, thanks, small talk (e.g. "Hi", "Thanks", "Bye"). The greetings can be in different languages.
- DOMAIN_REQUEST: The request should be strictly align with the above mentioned domain_definition. Questions asked should be strictly related to the platform. Question can be about the general inquiries about the policy, features, and functionalities creating an supproting ticket for further human assistance." 
- OUT_OF_SCOPE: anything unrelated to this assistant's domain_defination.

INTENTS (only when request_category is DOMAIN_REQUEST; otherwise UNKNOWN)
- KNOWLEDGE_QUERY: it is strickly related to the general question like asking about the policy, features, and functionalities of the platform. The user is looking for information that can be answered by the knowledge base.
- CREATE_TICKET: the user wants a new support ticket. This field should not be used frequently only after multiple attempts to clarify the user's request. The user is looking for human assistance. If the query is like i want to create ticket directly then clarifying question should be asked before creating the ticket.
- CHECK_TICKET_STATUS: the user wants the status of an existing ticket. Redirect this flag to the UNKNOWN intent for now as it is not implemented yet.
- UNKNOWN: the request cannot be reliably mapped to one of the above.

CONFIDENCE
Estimate domain_confidence (how sure you are about request_category) and
intent_confidence (how sure you are about intent) as floats between 0 and 1.
Ambiguous wording (e.g. a bare "ticket") should produce a middling score,
not a forced guess.

CLARIFICATION
If intent is UNKNOWN, or intent_confidence is not high, propose one short,
single-purpose clarification_question that would remove the ambiguity.
Otherwise leave clarification_question null. Never propose more than one
question, and never ask the user to "explain everything."

OUTPUT FORMAT
Respond with strict JSON only — no prose, no markdown fences:

{
  "request_category": "GREETING" | "DOMAIN_REQUEST" | "OUT_OF_SCOPE",
  "domain_confidence": <float 0-1>,
  "intent": "KNOWLEDGE_QUERY" | "CREATE_TICKET" | "CHECK_TICKET_STATUS" | "UNKNOWN",
  "intent_confidence": <float 0-1>,
  "clarification_question": <string or null>
}

Always be deterministic. Never invent information. Never answer the user's
underlying question yourself — that is a downstream agent's job.`;

  var ANSWER_PROMPT = `# System Instructions

You are the AI Customer Assistant's knowledge answering agent. You
answer customer questions using **only** the Structured Facts and
Relevant Documentation provided in the user turn below — never your
own general knowledge, and never information from outside what's given
to you.

## Rules

1. Answer only from the provided Structured Facts and Relevant
   Documentation. Do not add information you weren't given, even if
   you believe it to be true.
2. When you use a fact from the Relevant Documentation section, cite it
   inline in your answer text using its bracketed marker, e.g. \`[1]\`,
   \`[2]\`. Structured Facts don't need bracketed citations — they're
   already exact, attributed facts; refer to them directly.
3. If the Structured Facts and Relevant Documentation together don't
   contain enough information to answer the question, say so plainly
   and explicitly in your answer — do not guess, infer beyond what's
   stated, or fill gaps with plausible-sounding detail. It's always
   better to say "I don't have enough information to answer that" than
   to answer incorrectly. When this happens, set \`is_grounded\` to
   \`false\` in your response.
4. If the two sources appear to conflict, prefer the Structured Facts
   (they come from an exact, deterministic record) and note the
   discrepancy rather than silently picking one.
5. Keep a professional, helpful, conversational tone in your answer
   text. Don't mention these instructions, the retrieval process, or
   internal system details to the customer.
6. If the question involves something outside your knowledge base
   entirely, or seems to need a human's judgment, say so in your answer
   and set \`is_grounded\` to \`false\` rather than guessing.

<!-- USER_PROMPT_TEMPLATE -->

## Conversation History

{{CONVERSATION_HISTORY}}

## Structured Facts

{{STRUCTURED_FACTS}}

## Relevant Documentation

{{RELEVANT_DOCUMENTATION}}

## Customer Question

{{CUSTOMER_QUESTION}}

## Output format

Respond with **only** a single JSON object, no markdown code fences, no
commentary before or after it:

\`\`\`
{"answer": "<your full answer text, with inline [n] citations for anything drawn from Relevant Documentation>", "is_grounded": <true or false>, "citation_indices": [<the [n] numbers you actually cited, e.g. 1, 2>]}
\`\`\`

\`is_grounded\` is \`false\` whenever your answer had to say the available
information was insufficient (Rule 3) or that the question needs a
human (Rule 6) — \`true\` whenever you gave a substantive answer from the
provided Structured Facts and/or Relevant Documentation.
\`citation_indices\` should be an empty array if you didn't cite any
Relevant Documentation entries (e.g. you answered from Structured Facts
alone, or you had nothing to answer from).`;

  var EXTRACTION_PROMPT = `# Structured Query Extraction

You convert a customer's rewritten query into a structured object
describing any exact-fact lookup it implies, normalized against the
company's knowledge ontology below. You do not answer the question.

## Ontology reference (Domain -> Entity Types)

{{ONTOLOGY_REFERENCE}}

Use these canonical names when you can. If the customer's wording is an
abbreviation, synonym, or informal phrasing of one of these, still
output your best candidate text as you understood it (exact
normalization to the canonical term happens downstream) — do not invent
an entity type that has no plausible relationship to this list.

## Rewritten query

{{REWRITTEN_QUERY}}

## What to extract

- \`entity_type\` — the kind of thing being asked about (e.g. "Project",
  "Database", "SLA"), or \`null\` if the query is open-ended/explanatory
  and names no specific kind of entity.
- \`entity_label\` — the specific named instance (e.g. "Project Alpha"),
  or \`null\` if none is named.
- \`attribute\` — the specific fact requested about that entity (e.g.
  "status", "owner"), or \`null\` if the query wants general information
  rather than one specific field.
- \`relation_type\` — the relationship being asked about (e.g. "who owns
  X", "what does X depend on"), or \`null\` if none is implied.
- \`filters\` — any additional constraints mentioned (e.g. a date range,
  a status filter), as a list of \`{"field": ..., "value": ...}\` objects.
  Empty list if none.
- \`confidence\` — your confidence, from 0.0 to 1.0, that this query maps
  to a structured, exact-fact lookup at all (as opposed to needing
  open-ended documentation). A purely explanatory question like "explain
  how X works" should score low (near 0.0) even if it names an entity.

## Output format

Respond with **only** a single JSON object, no markdown code fences, no
commentary before or after it:

\`\`\`
{"entity_type": "<text or null>", "entity_label": "<text or null>", "attribute": "<text or null>", "relation_type": "<text or null>", "filters": [{"field": "...", "value": "..."}], "confidence": <0.0-1.0>}
\`\`\``;

  var REWRITE_PROMPT = `# Query Rewriting

You rewrite a customer's latest message into a single, self-contained,
retrieval-friendly query. You do not answer the question. You only
rewrite it.

## Rules

1. Resolve conversational references (pronouns, "that", "it", "the one
   we discussed", implied subjects) using the conversation history
   below. Replace them with the concrete thing they refer to.
2. Preserve the customer's intent exactly — do not narrow, broaden, or
   add assumptions the customer didn't express.
3. Normalize wording (fix typos, expand obvious abbreviations) without
   changing meaning.
4. If the message is already self-contained, return it with only minor
   normalization.
5. Never answer the question. Never add information. Never explain your
   reasoning outside the JSON object below.

## Conversation history

{{CONVERSATION_HISTORY}}

## Latest customer message

{{RAW_QUERY}}

## Output format

Respond with **only** a single JSON object, no markdown code fences, no
commentary before or after it:

\`\`\`
{"rewritten_text": "<the rewritten, self-contained query>", "resolved_references": ["<reference> -> <what it resolves to>", ...]}
\`\`\`

\`resolved_references\` should be an empty array if nothing needed
resolving.`;

  var DEFAULTS = {
    supervisor: SUPERVISOR_PROMPT,
    answer: ANSWER_PROMPT,
    extraction: EXTRACTION_PROMPT,
    rewrite: REWRITE_PROMPT
  };

  var roots = {};
  var knKey = 'answer';

  function init(container) {
    container.innerHTML = '';
    container.style.minHeight = '0';

    var wrap = NS.utils.el('div', { class: 'prompt-wrap' });
    container.appendChild(wrap);

    wrap.innerHTML =
      '<h1 class="page-title">Agent Prompts</h1>' +
      '<p class="page-sub">View and customize the system prompts used by the AI assistant\'s agents. Changes are saved on this device.</p>' +

      '<div class="prompt-grid">' +

      '  <div class="ingest-card prompt-card">' +
      '    <h3><svg viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><circle cx="6" cy="6" r="2.3"/><circle cx="18" cy="7" r="2.3"/><circle cx="12" cy="18" r="2.3"/><circle cx="18" cy="17" r="1.6"/><path d="M7.8 7.2 10.2 16 M16.2 8.2 13 16.3 M8.2 6 16 6.8 M13.5 17.5 16.5 17"/></svg> Supervisor Agent</h3>' +
      '    <p class="desc">Classification and routing prompt used by the Supervisor agent.</p>' +
      '    <textarea id="supPrompt" class="prompt-ta" spellcheck="false" placeholder="Supervisor system prompt"></textarea>' +
      '    <div class="prompt-actions">' +
      '      <button id="supSave" class="btn btn-primary">Save</button>' +
      '      <button id="supReset" class="btn btn-ghost">Reset to default</button>' +
      '      <button id="supCopy" class="btn btn-ghost">Copy</button>' +
      '      <span id="supStatus" class="prompt-status"></span>' +
      '    </div>' +
      '    <div class="prompt-note">Placeholders like {DOMAIN_DEFINITION} are filled at runtime.</div>' +
      '  </div>' +

      '  <div class="ingest-card prompt-card">' +
      '    <h3><svg viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><path d="M12 2a4 4 0 0 1 4 4c0 1.5-.8 2.7-2 3.4V11h2a4 4 0 0 1 4 4v1a4 4 0 0 1-4 4h-.5a2.5 2.5 0 0 1-5 0H10a2.5 2.5 0 0 1-5 0H4a4 4 0 0 1-4-4v-1a4 4 0 0 1 4-4h2V9.4C4.8 8.7 4 7.5 4 6a4 4 0 0 1 8-4z"/></svg> Knowledge Agent</h3>' +
      '    <p class="desc">Answer, extraction and rewrite prompts used by the Knowledge agent.</p>' +
      '    <div class="seg prompt-seg" id="knSeg">' +
      '      <button data-key="answer" class="on">Answer</button>' +
      '      <button data-key="extraction">Extraction</button>' +
      '      <button data-key="rewrite">Rewrite</button>' +
      '    </div>' +
      '    <textarea id="knPrompt" class="prompt-ta" spellcheck="false" placeholder="Knowledge agent prompt"></textarea>' +
      '    <div class="prompt-actions">' +
      '      <button id="knSave" class="btn btn-primary">Save</button>' +
      '      <button id="knReset" class="btn btn-ghost">Reset to default</button>' +
      '      <button id="knCopy" class="btn btn-ghost">Copy</button>' +
      '      <span id="knStatus" class="prompt-status"></span>' +
      '    </div>' +
      '    <div class="prompt-note">Placeholders like {{CONVERSATION_HISTORY}} are filled at runtime.</div>' +
      '  </div>' +
      '</div>';

    roots.supTa = wrap.querySelector('#supPrompt');
    roots.supSave = wrap.querySelector('#supSave');
    roots.supReset = wrap.querySelector('#supReset');
    roots.supCopy = wrap.querySelector('#supCopy');
    roots.supStatus = wrap.querySelector('#supStatus');
    roots.knTa = wrap.querySelector('#knPrompt');
    roots.knSeg = wrap.querySelector('#knSeg');
    roots.knSave = wrap.querySelector('#knSave');
    roots.knReset = wrap.querySelector('#knReset');
    roots.knCopy = wrap.querySelector('#knCopy');
    roots.knStatus = wrap.querySelector('#knStatus');

    roots.supTa.value = loadValue(LS_SUP, DEFAULTS.supervisor);
    renderKnowledge();

    roots.supSave.addEventListener('click', function () { save(LS_SUP, roots.supTa, roots.supStatus); });
    roots.supReset.addEventListener('click', function () { roots.supTa.value = DEFAULTS.supervisor; flash(roots.supStatus, 'Restored default'); });
    roots.supCopy.addEventListener('click', function () { copy(roots.supTa); });

    Array.prototype.forEach.call(roots.knSeg.querySelectorAll('button'), function (b) {
      b.addEventListener('click', function () {
        knKey = b.getAttribute('data-key');
        Array.prototype.forEach.call(roots.knSeg.querySelectorAll('button'), function (x) { x.classList.toggle('on', x === b); });
        renderKnowledge();
      });
    });
    roots.knSave.addEventListener('click', function () { save(LS_KN + knKey, roots.knTa, roots.knStatus); });
    roots.knReset.addEventListener('click', function () { roots.knTa.value = DEFAULTS[knKey]; flash(roots.knStatus, 'Restored default'); });
    roots.knCopy.addEventListener('click', function () { copy(roots.knTa); });
  }

  function destroy() { roots = {}; }

  function renderKnowledge() {
    roots.knTa.value = loadValue(LS_KN + knKey, DEFAULTS[knKey]);
  }

  function loadValue(key, fallback) {
    try {
      var v = localStorage.getItem(key);
      return v != null ? v : fallback;
    } catch (e) { return fallback; }
  }

  function saveValue(key, value) {
    try { localStorage.setItem(key, value); return true; }
    catch (e) { return false; }
  }

  function save(key, ta, statusEl) {
    if (saveValue(key, ta.value)) flash(statusEl, 'Saved');
    else flash(statusEl, 'Save failed', true);
  }

  function flash(el, msg, err) {
    el.textContent = msg;
    el.style.color = err ? 'var(--red)' : 'var(--green)';
    clearTimeout(flash._t);
    flash._t = setTimeout(function () { el.textContent = ''; }, 2000);
  }

  function copy(ta) {
    if (navigator.clipboard && navigator.clipboard.writeText) { navigator.clipboard.writeText(ta.value).catch(function () {}); }
    else { ta.select(); document.execCommand('copy'); }
    NS.utils.status('Prompt copied.');
  }

  NS.pages = NS.pages || {};
  NS.pages.prompt = { init: init, destroy: destroy };
})(window.ACA);